### ISSUE TITLE
Uncaught exception in Upstream upload function (with skipper-s3)

### ISSUE THREAD

**Issue by plato-cambrian** [balderdashy/skipper#88]
Hi, we're discussing an uncaught exception in [skipper-s3 #22](https://github.com/balderdashy/skipper-s3/issues/22) that kills the node process. This happens when skipper-s3 tries and fails to upload to s3 due to a temporary redirect. I suspect skipper may be involved due to the line `at  ... skipper\standalone\Upstream\prototype.fatalIncomingError.js:65:17`
```
events.js:74
        throw TypeError('Uncaught, unspecified "error" event.');
              ^
TypeError: Uncaught, unspecified "error" event.
    at TypeError (<anonymous>)
    at Writable.emit (events.js:74:15)
    at C:\dev\src\sopro\apps\chat\node_modules\skipper-s3\index.js:238:22
    at PassThrough.<anonymous> (C:\dev\src\sopro\apps\chat\node_modules\skipper-s3\node_modules\knox-mpu-alt\lib\multipartupload.js:328:16)
    at PassThrough.emit (events.js:117:20)
    at C:\dev\src\sopro\apps\chat\node_modules\skipper\standalone\Upstream\prototype.fatalIncomingError.js:65:17
    at Function.forEach (C:\dev\src\sopro\apps\chat\node_modules\skipper\node_modules\lodash\dist\lodash.js:3298:15)
    at lodash.ctor.(anonymous function) [as each] (C:\dev\src\sopro\apps\chat\node_modules\skipper\node_modules\lodash\dist\lodash.js:5925:31)
    at Upstream.fatalIncomingError (C:\dev\src\sopro\apps\chat\node_modules\skipper\standalone\Upstream\prototype.fatalIncomingError.js:55:18)
    at Writable.unableToUpload (C:\dev\src\sopro\apps\chat\node_modules\skipper\standalone\Upstream\prototype.upload.js:142:10)

    at Writable.g (events.js:180:16)
    at Writable.emit (events.js:117:20)
    at C:\dev\src\sopro\apps\chat\node_modules\skipper-s3\index.js:238:22
    at C:\dev\src\sopro\apps\chat\node_modules\skipper-s3\node_modules\knox-mpu-alt\lib\multipartupload.js:217:42
    at ClientRequest.<anonymous> (C:\dev\src\sopro\apps\chat\node_modules\skipper-s3\node_modules\knox-mpu-alt\lib\multipartupload.js:129:25)
    at ClientRequest.emit (events.js:95:17)
```

**Comment by mattlo** [comment]
@plato-cambrian Ran into this issue as well. To catch theses errors, there's a few ways you can do it:
- Fork a cluster and invoke the process within
- Use experimental `domain` to isolate the error
- `process.on('uncaughtException')` <- not really an option

I used `domain`, its straight forward implementing and seemed to be a leading solution for future error handling around foreign dependencies. Of course its still experimental but will get the job done in catching the error and propagating the *actual* s3 error message (instead of uncommenting logs as mentioned by https://github.com/balderdashy/skipper-s3/issues/22#issuecomment-107696944. I ran quite a few unit tests to try to break `domain` (even tests outside our implementation usage) and it appeared to work as intended. I can't recommend to use it on production, but based on the small data I gathered, it's enough for me to ship it. I'll wrap it within a cluster if I run into issues later on.

**Comment by sailsbot** [comment]
Thanks for posting, @plato-cambrian.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
